tests/PHPStan/Rules/Classes/AllowedSubTypesRuleTest.php

```diff
@@ -0,0 +1,38 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Classes;
+
+use PHPStan\Rules\Rule;
+use PHPStan\Testing\RuleTestCase;
+
+/**
+ * @extends RuleTestCase<AllowedSubTypesRule>
+ */
+class AllowedSubTypesRuleTest extends RuleTestCase
+{
+
+	protected function getRule(): Rule
+	{
+		return new AllowedSubTypesRule();
+	}
+
+	public function testRule(): void
+	{
+		require __DIR__ . '/data/allowed-sub-types.php';
+		$this->analyse([__DIR__ . '/data/allowed-sub-types.php'], [
```

Comment:
Can you also test that:

1) `DateTime` CAN be extended by userland classes
2) `DateTimeInterface` CANNOT be implemented by userland classes

https://3v4l.org/f7V26

Reply:
Good point, while we're at it, I've also added an extension (and test) for `Throwable`

Reply:
Hm. Turns out you _can_ add an explicit `implements DateTimeInterface` to a class as long as it is also a descendant of `DateTime` or `DateTimeImmutable` (https://3v4l.org/nOU76). The same goes for `Throwable`.

I could account for that in the rule, but I'm afraid it might not work that well for other use cases – including my original use case of sealed classes.

I'm starting to think this is trying to accomplish two different things, and perhaps we should only focus on one here – provide better type narrowing for restricted hierarchies. I guess we could come up with better names to make it clear that the extension only focuses on this and that it alone doesn't give any guarantees that the inheritance hierarchy is actually restricted, leaving it up to the applications (and phpstan-extensions) to enforce that via a custom rule based on their specific needs and conventions.

Having said that, I think that PHP natives like `DateTimeInterface` and `Throwable` _should_ be covered by PHPStan, but that a more targeted rule which takes the aforementioned quirks into account would be a better fit in this case.

WDYT?

Reply:
I think that could be doable. But there are other edge cases, such as https://3v4l.org/nbLp6

Reply:
I think we have two possible paths to take here:

1) Include the new rule in this PR but continue hardcoding sealed DateTimeInterface class hierarchy as it is now on 1.8.x. So DateTimeInterfaceAllowedSubTypesClassReflectionExtension and ThrowableAllowedSubTypesClassReflectionExtension wouldn't be part of this PR. This means that the new rule would only target custom allowedSubTypes extensions.
2) Leave things as they are now in this PR but remove AllowedSubTypesRule.

I think 1) makes more sense because people aren't required to implement their custom (very repetetive) rules, they just have to provide custom extension to define the class hierarchy. The downside is that PHPStan internals will be a bit more ugly but that's fine.

WDYT? After this we can implement a completely separate rule to detect situations when the user shouldn't implement DateTimeInterface, Throwable etc...

Reply:
Okay, that makes sense to me :)